GH-48220: [C++] Skip arrow-buffer & arrow-misc testcase runs on s390x

Vishwanatha-HD opened the pull request:
### Rationale for this change

This PR is intended to enable Parquet DB support on Big-endian (s390x) systems. The fix in this PR fixes the "arrow-buffer-test" & "arrow-misc-test" failures on s390x.

### What changes are included in this PR?

The fix includes changes to following files:
cpp/src/arrow/buffer_test.cc
cpp/src/arrow/memory_pool_test.cc

### Are these changes tested?

Yes. The changes are tested on s390x arch to make sure things are working fine. The fix is also tested on x86 arch, to make sure there is no new regression introduced.

### Are there any user-facing changes?

No.

GitHub main Issue link: https://github.com/apache/arrow/issues/48151
* GitHub Issue: #48220

Vishwanatha-HD committed:
GH-48220 Fix arrow-buffer & arrow-misc testcase failures to enable Parquet DB support on s390x

kou commented:
Could you update the PR title and description?
* This doesn't fix failures. This just skips failed tests.
* This isn't for Parquet. This is for Arrow.

kiszk commented on cpp/src/arrow/memory_pool_test.cc:
```diff
@@ -68,8 +69,9 @@ TYPED_TEST_SUITE_P(TestMemoryPool);
 
 TYPED_TEST_P(TestMemoryPool, MemoryTracking) { this->TestMemoryTracking(); }
 
+// Skip this test on big-endian architectures (e.g., s390x)
```
ditto

Vishwanatha-HD replied:
These are my observations:
1) The "mimalloc" allocation request enters mimalloc's large-object allocation path.
2) Mimalloc attempts to calculate reservation sizes, alignment, segments, and OS mappings.
3) On s390x systems with huge virtual address spaces and Linux overcommit enabled, the allocation may not fail immediately.
4) Mimalloc is spending a significant amount of time trying to satisfy or validate the request before ultimately failing.
5) The testcase "hangs" because the control never returning to Arrow.

kiszk replied:
ditto

Vishwanatha-HD commented:
@pitrou @kou @kiszk @zanmato1984
Hi All,
I know its been long time since I have my PRs opened.. Can you please help me with review and merging to the upstream. Is there anything more that you people want me to do it, I would be happy to work on it. Please suggest.
I have verified that with my fix all the 113 testcase passes without any issues.
Thanks.

Vishwanatha-HD commented on cpp/src/arrow/buffer_test.cc:
```diff
@@ -996,7 +997,8 @@ TYPED_TEST(TypedTestBuffer, TypedResize) {
 // This test doesn't play nice with AddressSanitizer
-#ifndef ADDRESS_SANITIZER
+// Skip this test on big-endian architectures (e.g., s390x)
+#if !defined(ADDRESS_SANITIZER) && ARROW_LITTLE_ENDIAN
```
These are my observations:
1) The "mimalloc" allocation request enters mimalloc's large-object allocation path.
2) Mimalloc attempts to calculate reservation sizes, alignment, segments, and OS mappings.
3) On s390x systems with huge virtual address spaces and Linux overcommit enabled, the allocation may not fail immediately.
4) Mimalloc is spending a significant amount of time trying to satisfy or validate the request before ultimately failing.
5) The testcase "hangs" because the control never returning to Arrow.

kiszk replied:
At least, such a justification should be written in the source code.

Vishwanatha-HD reviewed (commented):
I have tried answering your questions. Please take a look. Thanks.

pitrou commented:
@Vishwanatha-HD Can you try rebasing your PR? We are using a newer mimalloc version now and your test  skips could have become unnecessary.

Vishwanatha-HD commented:
Hi @pitrou.. I did the rebase of my PR.. 
I tested with the newer mimalloc and the testcase still hangs if I dont skip it.. 

ctest -R "buffer|memory_pool" -VV
UpdateCTestConfiguration  from :arrow/cpp/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :arrow/cpp/build/DartConfiguration.tcl
Test project /home/vishwa/golang/arrowGitRepo/arrow/cpp/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 22
    Start 22: arrow-buffer-test

22: Test command: arrow/cpp/build-support/run-test.sh "
arrowGitRepo/arrow/cpp/build" "test" "arrowGitRepo/arrow/cpp/build/debug//arrow-buffer-test"
22: Working Directory: arrowGitRepo/arrow/cpp/build/src/arrow
22: Test timeout computed to be: 10000000
22: Running arrow-buffer-test, redirecting output into arrowGitRepo/arrow/cpp/build/build/test-logs/arrow-buffer-test.txt (attempt 1/1)
......................
......................
Never comes out of this.. 

To confirm the working, I did set the "ARROW_DEFAULT_MEMORY_POOL" to "system" and ran the test again.. I see that all the tests passes in this case.. 
export ARROW_DEFAULT_MEMORY_POOL=system

ctest -R "buffer|memory_pool" -VV
UpdateCTestConfiguration  from :/home/vishwa/golang/arrowGitRepo/arrow/cpp/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/vishwa/golang/arrowGitRepo/arrow/cpp/build/DartConfiguration.tcl
Test project /home/vishwa/golang/arrowGitRepo/arrow/cpp/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 22
    Start 22: arrow-buffer-test

22: Test command: /home/vishwa/golang/arrowGitRepo/arrow/cpp/build-support/run-test.sh "/home/vishwa/golang/arrowGitRepo/arrow/cpp/build" "test" "/home/vishwa/golang/arrowGitRepo/arrow/cpp/build/debug//arrow-buffer-test"
22: Working Directory: /home/vishwa/golang/arrowGitRepo/arrow/cpp/build/src/arrow
22: Test timeout computed to be: 10000000
22: Running arrow-buffer-test, redirecting output into /home/vishwa/golang/arrowGitRepo/arrow/cpp/build/build/test-logs/arrow-buffer-test.txt (attempt 1/1)
22: Running main() from gmock_main.cc
22: [==========] Running 47 tests from 15 test suites.
22: [----------] Global test environment set-up.
22: [----------] 3 tests from TestDevice
22: [ RUN      ] TestDevice.Basics
22: [       OK ] TestDevice.Basics (0 ms)
22: [ RUN      ] TestDevice.Copy
22: [       OK ] TestDevice.Copy (0 ms)
22: [ RUN      ] TestDevice.View
22: [       OK ] TestDevice.View (0 ms)
22: [----------] 3 tests from TestDevice (0 ms total)
22: 
22: [----------] 3 tests from TestAllocate
22: [ RUN      ] TestAllocate.Basics
22: [       OK ] TestAllocate.Basics (0 ms)
22: [ RUN      ] TestAllocate.Bitmap
22: [       OK ] TestAllocate.Bitmap (0 ms)
22: [ RUN      ] TestAllocate.EmptyBitmap
22: [       OK ] TestAllocate.EmptyBitmap (0 ms)
22: [----------] 3 tests from TestAllocate (0 ms total)
22: 
22: [----------] 14 tests from TestBuffer
22: [ RUN      ] TestBuffer.FromStdString
22: [       OK ] TestBuffer.FromStdString (0 ms)
22: [ RUN      ] TestBuffer.Alignment
22: [       OK ] TestBuffer.Alignment (0 ms)
22: [ RUN      ] TestBuffer.FromStdStringWithMemory
22: [       OK ] TestBuffer.FromStdStringWithMemory (0 ms)
22: [ RUN      ] TestBuffer.EqualsWithSameContent
22: [       OK ] TestBuffer.EqualsWithSameContent (0 ms)
22: [ RUN      ] TestBuffer.EqualsWithSameBuffer
22: [       OK ] TestBuffer.EqualsWithSameBuffer (0 ms)
22: [ RUN      ] TestBuffer.CopySlice
22: [       OK ] TestBuffer.CopySlice (0 ms)
22: [ RUN      ] TestBuffer.CopySliceEmpty
22: [       OK ] TestBuffer.CopySliceEmpty (0 ms)
22: [ RUN      ] TestBuffer.ToHexString
22: [       OK ] TestBuffer.ToHexString (0 ms)
22: [ RUN      ] TestBuffer.SliceBuffer
22: [       OK ] TestBuffer.SliceBuffer (0 ms)
22: [ RUN      ] TestBuffer.SliceBufferSafe
22: [       OK ] TestBuffer.SliceBufferSafe (0 ms)
22: [ RUN      ] TestBuffer.FromStringRvalue
22: [       OK ] TestBuffer.FromStringRvalue (0 ms)
22: [ RUN      ] TestBuffer.SliceMutableBuffer
22: [       OK ] TestBuffer.SliceMutableBuffer (0 ms)
22: [ RUN      ] TestBuffer.GetReader
22: [       OK ] TestBuffer.GetReader (0 ms)
22: [ RUN      ] TestBuffer.GetWriter
22: [       OK ] TestBuffer.GetWriter (0 ms)
22: [----------] 14 tests from TestBuffer (0 ms total)
22: 
22: [----------] 1 test from TestMutableBuffer
22: [ RUN      ] TestMutableBuffer.Wrap
22: [       OK ] TestMutableBuffer.Wrap (0 ms)
22: [----------] 1 test from TestMutableBuffer (0 ms total)
22: 
22: [----------] 1 test from TestAllocateBuffer
22: [ RUN      ] TestAllocateBuffer.ZeroSize
22: [       OK ] TestAllocateBuffer.ZeroSize (0 ms)
22: [----------] 1 test from TestAllocateBuffer (0 ms total)
22: 
22: [----------] 2 tests from TestAllocateResizableBuffer
22: [ RUN      ] TestAllocateResizableBuffer.ZeroSize
22: [       OK ] TestAllocateResizableBuffer.ZeroSize (0 ms)
22: [ RUN      ] TestAllocateResizableBuffer.ZeroResize
22: [       OK ] TestAllocateResizableBuffer.ZeroResize (0 ms)
22: [----------] 2 tests from TestAllocateResizableBuffer (0 ms total)
22: 
22: [----------] 4 tests from TestBufferBuilder
22: [ RUN      ] TestBufferBuilder.ResizeReserve
22: [       OK ] TestBufferBuilder.ResizeReserve (0 ms)
22: [ RUN      ] TestBufferBuilder.Alignment
22: [       OK ] TestBufferBuilder.Alignment (0 ms)
22: [ RUN      ] TestBufferBuilder.Finish
22: [       OK ] TestBufferBuilder.Finish (0 ms)
22: [ RUN      ] TestBufferBuilder.FinishEmpty
22: [       OK ] TestBufferBuilder.FinishEmpty (0 ms)
22: [----------] 4 tests from TestBufferBuilder (0 ms total)
22: 
22: [----------] 2 tests from TypedTestBufferBuilder/0, where TypeParam = short
22: [ RUN      ] TypedTestBufferBuilder/0.BasicTypedBufferBuilderUsage
22: [       OK ] TypedTestBufferBuilder/0.BasicTypedBufferBuilderUsage (0 ms)
22: [ RUN      ] TypedTestBufferBuilder/0.AppendCopies
22: [       OK ] TypedTestBufferBuilder/0.AppendCopies (0 ms)
22: [----------] 2 tests from TypedTestBufferBuilder/0 (0 ms total)
22: 
22: [----------] 2 tests from TypedTestBufferBuilder/1, where TypeParam = unsigned int
22: [ RUN      ] TypedTestBufferBuilder/1.BasicTypedBufferBuilderUsage
22: [       OK ] TypedTestBufferBuilder/1.BasicTypedBufferBuilderUsage (0 ms)
22: [ RUN      ] TypedTestBufferBuilder/1.AppendCopies
22: [       OK ] TypedTestBufferBuilder/1.AppendCopies (0 ms)
22: [----------] 2 tests from TypedTestBufferBuilder/1 (0 ms total)
22: 
22: [----------] 2 tests from TypedTestBufferBuilder/2, where TypeParam = double
22: [ RUN      ] TypedTestBufferBuilder/2.BasicTypedBufferBuilderUsage
22: [       OK ] TypedTestBufferBuilder/2.BasicTypedBufferBuilderUsage (0 ms)
22: [ RUN      ] TypedTestBufferBuilder/2.AppendCopies
22: [       OK ] TypedTestBufferBuilder/2.AppendCopies (0 ms)
22: [----------] 2 tests from TypedTestBufferBuilder/2 (0 ms total)
22: 
22: [----------] 3 tests from TestBoolBufferBuilder
22: [ RUN      ] TestBoolBufferBuilder.Basics
22: [       OK ] TestBoolBufferBuilder.Basics (0 ms)
22: [ RUN      ] TestBoolBufferBuilder.AppendCopies
22: [       OK ] TestBoolBufferBuilder.AppendCopies (0 ms)
22: [ RUN      ] TestBoolBufferBuilder.Reserve
22: [       OK ] TestBoolBufferBuilder.Reserve (0 ms)
22: [----------] 3 tests from TestBoolBufferBuilder (0 ms total)
22: 
22: [----------] 4 tests from TypedTestBuffer/0, where TypeParam = std::shared_ptr<arrow::ResizableBuffer>
22: [ RUN      ] TypedTestBuffer/0.IsMutableFlag
22: [       OK ] TypedTestBuffer/0.IsMutableFlag (0 ms)
22: [ RUN      ] TypedTestBuffer/0.Resize
22: [       OK ] TypedTestBuffer/0.Resize (0 ms)
22: [ RUN      ] TypedTestBuffer/0.TypedResize
22: [       OK ] TypedTestBuffer/0.TypedResize (0 ms)
22: [ RUN      ] TypedTestBuffer/0.ResizeOOM
22: [       OK ] TypedTestBuffer/0.ResizeOOM (0 ms)
22: [----------] 4 tests from TypedTestBuffer/0 (0 ms total)
22: 
22: [----------] 4 tests from TypedTestBuffer/1, where TypeParam = std::unique_ptr<arrow::ResizableBuffer,std::default_delete<arrow::ResizableBuffer> >
22: [ RUN      ] TypedTestBuffer/1.IsMutableFlag
22: [       OK ] TypedTestBuffer/1.IsMutableFlag (0 ms)
22: [ RUN      ] TypedTestBuffer/1.Resize
22: [       OK ] TypedTestBuffer/1.Resize (0 ms)
22: [ RUN      ] TypedTestBuffer/1.TypedResize
22: [       OK ] TypedTestBuffer/1.TypedResize (0 ms)
22: [ RUN      ] TypedTestBuffer/1.ResizeOOM
22: [       OK ] TypedTestBuffer/1.ResizeOOM (0 ms)
22: [----------] 4 tests from TypedTestBuffer/1 (0 ms total)
22: 
22: [----------] 1 test from TestBufferConcatenation
22: [ RUN      ] TestBufferConcatenation.EmptyBuffer
22: [       OK ] TestBufferConcatenation.EmptyBuffer (0 ms)
22: [----------] 1 test from TestBufferConcatenation (0 ms total)
22: 
22: [----------] 1 test from TestDeviceRegistry
22: [ RUN      ] TestDeviceRegistry.Basics
22: [       OK ] TestDeviceRegistry.Basics (0 ms)
22: [----------] 1 test from TestDeviceRegistry (0 ms total)
22: 
22: [----------] Global test environment tear-down
22: [==========] 47 tests from 15 test suites ran. (1 ms total)
22: [  PASSED  ] 47 tests.
22: /home/vishwa/golang/arrowGitRepo/arrow/cpp/build/src/arrow
1/2 Test #22: arrow-buffer-test ................   Passed    0.06 sec
test 56
    Start 56: arrow-io-buffered-test

56: Test command: /home/vishwa/golang/arrowGitRepo/arrow/cpp/build-support/run-test.sh "/home/vishwa/golang/arrowGitRepo/arrow/cpp/build" "test" "/home/vishwa/golang/arrowGitRepo/arrow/cpp/build/debug//arrow-io-buffered-test"
56: Working Directory: /home/vishwa/golang/arrowGitRepo/arrow/cpp/build/src/arrow/io
56: Test timeout computed to be: 10000000
56: Running arrow-io-buffered-test, redirecting output into /home/vishwa/golang/arrowGitRepo/arrow/cpp/build/build/test-logs/arrow-io-buffered-test.txt (attempt 1/1)
56: Running main() from gmock_main.cc
56: [==========] Running 34 tests from 5 test suites.
56: [----------] Global test environment set-up.
56: [----------] 12 tests from TestBufferedOutputStream
56: [ RUN      ] TestBufferedOutputStream.DestructorClosesFile
56: [       OK ] TestBufferedOutputStream.DestructorClosesFile (0 ms)
56: [ RUN      ] TestBufferedOutputStream.Detach
56: [       OK ] TestBufferedOutputStream.Detach (0 ms)
56: [ RUN      ] TestBufferedOutputStream.ExplicitCloseClosesFile
56: [       OK ] TestBufferedOutputStream.ExplicitCloseClosesFile (0 ms)
56: [ RUN      ] TestBufferedOutputStream.InvalidWrites
56: [       OK ] TestBufferedOutputStream.InvalidWrites (0 ms)
56: [ RUN      ] TestBufferedOutputStream.TinyWrites
56: [       OK ] TestBufferedOutputStream.TinyWrites (0 ms)
56: [ RUN      ] TestBufferedOutputStream.SmallWrites
56: [       OK ] TestBufferedOutputStream.SmallWrites (7 ms)
56: [ RUN      ] TestBufferedOutputStream.MixedWrites
56: [       OK ] TestBufferedOutputStream.MixedWrites (3 ms)
56: [ RUN      ] TestBufferedOutputStream.LargeWrites
56: [       OK ] TestBufferedOutputStream.LargeWrites (10 ms)
56: [ RUN      ] TestBufferedOutputStream.Flush
56: [       OK ] TestBufferedOutputStream.Flush (0 ms)
56: [ RUN      ] TestBufferedOutputStream.SetBufferSize
56: [       OK ] TestBufferedOutputStream.SetBufferSize (0 ms)
56: [ RUN      ] TestBufferedOutputStream.Tell
56: [       OK ] TestBufferedOutputStream.Tell (0 ms)
56: [ RUN      ] TestBufferedOutputStream.TruncatesFile
56: [       OK ] TestBufferedOutputStream.TruncatesFile (0 ms)
56: [----------] 12 tests from TestBufferedOutputStream (22 ms total)
56: 
56: [----------] 8 tests from TestBufferedInputStream
56: [ RUN      ] TestBufferedInputStream.InvalidReads
56: [       OK ] TestBufferedInputStream.InvalidReads (0 ms)
56: [ RUN      ] TestBufferedInputStream.BasicOperation
56: [       OK ] TestBufferedInputStream.BasicOperation (0 ms)
56: [ RUN      ] TestBufferedInputStream.Detach
56: [       OK ] TestBufferedInputStream.Detach (0 ms)
56: [ RUN      ] TestBufferedInputStream.ReadBuffer
56: [       OK ] TestBufferedInputStream.ReadBuffer (0 ms)
56: [ RUN      ] TestBufferedInputStream.SetBufferSize
56: [       OK ] TestBufferedInputStream.SetBufferSize (0 ms)
56: [ RUN      ] TestBufferedInputStream.BufferSizeLimit
56: [       OK ] TestBufferedInputStream.BufferSizeLimit (0 ms)
56: [ RUN      ] TestBufferedInputStream.PeekPastBufferedBytes
56: [       OK ] TestBufferedInputStream.PeekPastBufferedBytes (0 ms)
56: [ RUN      ] TestBufferedInputStream.PeekAfterExhaustingBuffer
56: [       OK ] TestBufferedInputStream.PeekAfterExhaustingBuffer (0 ms)
56: [----------] 8 tests from TestBufferedInputStream (1 ms total)
56: 
56: [----------] 5 tests from TestBufferedInputStreamBound
56: [ RUN      ] TestBufferedInputStreamBound.Basics
56: [       OK ] TestBufferedInputStreamBound.Basics (0 ms)
56: [ RUN      ] TestBufferedInputStreamBound.LargeFirstPeek
56: [       OK ] TestBufferedInputStreamBound.LargeFirstPeek (0 ms)
56: [ RUN      ] TestBufferedInputStreamBound.UnboundedPeek
56: [       OK ] TestBufferedInputStreamBound.UnboundedPeek (0 ms)
56: [ RUN      ] TestBufferedInputStreamBound.OneByteReads
56: [       OK ] TestBufferedInputStreamBound.OneByteReads (0 ms)
56: [ RUN      ] TestBufferedInputStreamBound.BufferExactlyExhausted
56: [       OK ] TestBufferedInputStreamBound.BufferExactlyExhausted (0 ms)
56: [----------] 5 tests from TestBufferedInputStreamBound (0 ms total)
56: 
56: [----------] 7 tests from TestBufferedInputStreamChunk
56: [ RUN      ] TestBufferedInputStreamChunk.NoRead
56: [       OK ] TestBufferedInputStreamChunk.NoRead (0 ms)
56: [ RUN      ] TestBufferedInputStreamChunk.LargeRead
56: [       OK ] TestBufferedInputStreamChunk.LargeRead (0 ms)
56: [ RUN      ] TestBufferedInputStreamChunk.SmallReadThenLargeRead
56: [       OK ] TestBufferedInputStreamChunk.SmallReadThenLargeRead (0 ms)
56: [ RUN      ] TestBufferedInputStreamChunk.BufferWholeChunk
56: [       OK ] TestBufferedInputStreamChunk.BufferWholeChunk (0 ms)
56: [ RUN      ] TestBufferedInputStreamChunk.BufferLargerThanFileSize
56: [       OK ] TestBufferedInputStreamChunk.BufferLargerThanFileSize (0 ms)
56: [ RUN      ] TestBufferedInputStreamChunk.BufferLargerThanReadBound
56: [       OK ] TestBufferedInputStreamChunk.BufferLargerThanReadBound (0 ms)
56: [ RUN      ] TestBufferedInputStreamChunk.BufferSmallerThanReadBound
56: [       OK ] TestBufferedInputStreamChunk.BufferSmallerThanReadBound (0 ms)
56: [----------] 7 tests from TestBufferedInputStreamChunk (0 ms total)
56: 
56: [----------] 2 tests from TestBufferedInputStreamRandom
56: [ RUN      ] TestBufferedInputStreamRandom.ReadsWithoutReadBound
56: [       OK ] TestBufferedInputStreamRandom.ReadsWithoutReadBound (4 ms)
56: [ RUN      ] TestBufferedInputStreamRandom.ReadsWithReadBound
56: [       OK ] TestBufferedInputStreamRandom.ReadsWithReadBound (2 ms)
56: [----------] 2 tests from TestBufferedInputStreamRandom (6 ms total)
56: 
56: [----------] Global test environment tear-down
56: [==========] 34 tests from 5 test suites ran. (32 ms total)
56: [  PASSED  ] 34 tests.
56: /home/vishwa/golang/arrowGitRepo/arrow/cpp/build/src/arrow/io
2/2 Test #56: arrow-io-buffered-test ...........   Passed    0.08 sec

The following tests passed:
	arrow-buffer-test
	arrow-io-buffered-test

100% tests passed, 0 tests failed out of 2

Label Time Summary:
arrow-tests    =   0.14 sec*proc (2 tests)
unittest       =   0.14 sec*proc (2 tests)

Total Test time (real) =   0.15 sec

Vishwanatha-HD commented:
Hi @kou, @pitrou .
I have rebased this PR onto the main. Please check.. Thanks..
